pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java

```diff
@@ -0,0 +1,383 @@
+
+                return future;
+            } else if (status != Status.Enabled && shouldBeEnabled) {
+                // Enable deduping
+                CompletableFuture<Void> future = new CompletableFuture<>();
+                managedLedger.asyncOpenCursor(PersistentTopic.DEDUPLICATION_CURSOR_NAME, new OpenCursorCallback() {
+
+                    @Override
+                    public void openCursorComplete(ManagedCursor cursor, Object ctx) {
+                        // We don't want to retain cache for this cursor
+                        cursor.setInactive();
+                        managedCursor = cursor;
+                        recoverSequenceIdsMap().thenRun(() -> {
+                            status = Status.Enabled;
+                            future.complete(null);
+                            log.info("[{}] Enabled deduplication", topic.getName());
+                        }).exceptionally(ex -> {
+                            status = Status.Failed;
+                            log.warn("[{}] Failed to enable deduplication: {}", topic.getName(), ex.getMessage());
+                            future.completeExceptionally(ex);
+                            return null;
+                        });
+                    }
+
+                    @Override
+                    public void openCursorFailed(ManagedLedgerException exception, Object ctx) {
+                        log.warn("[{}] Failed to enable deduplication: {}", topic.getName(), exception.getMessage());
+                        future.completeExceptionally(exception);
+                    }
+
+                }, null);
+                return future;
+            } else {
+                // Nothing to do, we are in the correct state
+                return CompletableFuture.completedFuture(null);
+            }
+        });
+    }
+
+    public boolean isEnabled() {
+        return status == Status.Enabled;
+    }
+
+    /**
+     * Assess whether the message was already stored in the topic
+     */
+    public boolean isMessageAlreadyStored(String producerName, long sequenceId) {
+        if (!isEnabled()) {
+            return false;
+        }
+
+        // Synchronize the get() and subsequent put() on the map. This would only be relevant if the producer
+        // disconnects and re-connects very quickly. At that point the call can be coming from a different thread
+        synchronized (highestSequencedPushed) {
+            Long lastSequenceIdPushed = highestSequencedPushed.get(producerName);
+            if (lastSequenceIdPushed != null && sequenceId <= lastSequenceIdPushed) {
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] Message identified as duplicated producer={} seq-id={} -- highest-seq-id={}",
+                            topic.getName(), producerName, sequenceId, lastSequenceIdPushed);
+                }
+                return true;
+            }
+
+            highestSequencedPushed.put(producerName, sequenceId);
```

Comment:
I feel `isMessageAlreadyStored` name is little confusing.. as per method name it should not update the map. I know we want to add the sequenceId as well. how about `pushSequenceId()` and if it returns `false` means couldn't store.

Reply:
Ok, let me rethink the method name.

Reply:
Renaming into `shouldPublishNextMessage()`